GSA-TTS/FAC
# Fix submission race condition

*We race to the end*
_Auditee and auditee_
_Both win, we all lose_ 

The submission process is susceptible to races. In particular, it was possible for a user to press the final submission button twice, triggering two (or more) submission processes that would then [race](https://en.wikipedia.org/wiki/Race_condition) to finalize a submission and data dissemination.

This PR addresses https://github.com/GSA-TTS/FAC/issues/4272 in terms of fixing the race. It does not fix the stuck audit, and it does not disable the submission button (that is in https://github.com/GSA-TTS/FAC/issues/4274). The fix now guarantees that only one submission runs to completion; if a second submission is made, no error is thrown, an INFO statement is logged to NR, and the user is quietly brought back to the default landing page. (If they managed to press the submission button twice, they would still see an end result of a completely submitted audit; this feels like appropriate behavior.)

A thorough README is included in the PR that documents the race condition as well as the fix. It is included in the hopes that future developers will not undo protections that prevent the race from occurring.

A new compose is included that stands up two separate instances of the FAC. This was used to force the race hazard across FAC instances (FAC1 and FAC2), simulating our cloud.gov environment. It might be removed at some future point. To test thoroughly, sleep statements need to be inserted into the code, which is demonstrated in the penultimate commit; the approach was reviewed with and demonstrated to @sambodeme as part of the final development.

E2E tests passed locally. Unit tests passed locally.

Local manual testing
- Start up using `docker-compose -f docker-compose-dual-fac.yml up`. You should be able to reach the app from http://localhost:8000/ and http://localhost:8001/.
- Add the following method to [SubmissionView](https://github.com/GSA-TTS/FAC/blob/b3cf5a62cce870d2f1c146397c715d4cfa9ce5e8/backend/audit/views.py#L270) and call it from one of the example locations in [this commit](https://github.com/GSA-TTS/FAC/pull/4292/commits/1cc9170b48b005315f5caebc64b57d0aa770f839)
  - ```
    def simulate_delay(self, instance, duration):
        running_version = os.getenv("FAC_INSTANCE_NAME")
        if running_version == instance:
            logger.info("{} sleeping for {}s".format(running_version, duration))
            time.sleep(duration)
        else:
            logger.info("{} continuing...".format(running_version))
    ```
- Have an audit ready to submit in 2 tabs. You can set this up by commenting out all the steps after `// Submit` in `backend/cypress/support/full-submission.js` and running the Cypress tests. Make sure the auditee account is one you're able to log into locally for the next step, but it also must be different from the auditor account.
- While logged in as the auditee, go to the submission step for the audit in two tabs, one for 8000 and one for 8001
- Submit the audit on 8000 first, immediately followed by 8001
- 8001 will just redirect you to the audits home page
- 8000 will delay for 20s and eventually submit/disseminate successfully
- If you want to retest on the same audit, you'll have to use DBeaver to TRUNCATE CASCADE the dissemination tables, and also manually roll back the submission from "disseminated" to "auditee_certified" (as well as remove the last two list elements from the transition list/timestamps.)

## PR Checklist: Submitter


## PR Checklist: Reviewer


The larger the PR, the stricter we should be about these points.

## Pre Merge Checklist: Merger
